WIP: update README.md (#59)
rust-vmm/community: README.md

# Community

## What is rust-vmm

> rust-vmm is an open-source project that enables the community to build custom

Reviewer: s/that enables/enabling/

Virtual Machine Monitors (VMMs) and Hypervisors. It offers a base of

> virtualization packages that other projects can use for rapidly developing

Reviewer: s/other projects/any project/
s/use for rapidly developing/benefit from in order to quickly develop/

virtualization solutions while focusing on the key differentiators of their

> product rather than re-implementing common components like KVM wrappers, Virtio

Reviewer: s/Virtio/virtio/

devices and other VMM libraries.

> The rust-vmm project is organized as a shared-effort, shared-ownership

Reviewer: s/,/ and/

open-source project that includes (so far) contributors from Alibaba, AWS,
Cloud Base, Crowdstrike, Intel, Google, Red Hat as well as individual
contributors.

> Each virtualization component lies in a separated GitHub repository under the

Reviewer: s/component/package/ (for consistency with above)
s/lies/lives/

Author: modified the above mention with component.

rust-vmm organization. The code is organized such that one repository

> corresponds to a

Reviewer: s/a/a single/

Author: replaced with one.

[Rust crate](https://doc.rust-lang.org/book/ch07-01-packages-and-crates-for-making-libraries-and-executables.html).

## Why rust-vmm

- **Reduce code duplication**. The initial thought with rust-vmm was to create a
  place for sharing common virtualization components between two existing VMMs
  written in Rust:
  [CrosVM](https://chromium.googlesource.com/chromiumos/platform/crosvm/) and
  [Firecracker](https://github.com/firecracker-microvm/firecracker/). These

>   two projects have similar code for calling KVM ioctls, working with the

Reviewer: s/working with/managing/

>   Virtual Machine memory, virtio implementation and others. TODO.

Reviewer: s/virtio/interacting with virtio/

> - **Faster development for new custom VMMs**. rust-vmm provides a base of

Reviewer: s/Faster development for new custom VMMs/Faster development/

>   virtualization components which are meant to be generic enough to be used by

Reviewer: I think we should be consistent in the whole document regarding virtualization "components" or "packages".
s/generic enough to be used by/generic to be consumed from/

Author: I tried to use them interchangably so we don't end up having the word component many times in one phrase. But I can also just use component everywhere.

>   other projects besides CrosVM and Firecracker. One example that often comes

Reviewer: s/often comes up/is often mentioned/

>   up is building a container specific VMM. In doing so, the container VMM can

Reviewer: s/In doing/By doing/
s/the container VMM can/this hypothetical container VMM could/

>   reuse most of the rust-vmm components and just build the glue around these

Reviewer: s/just/simply/
s/around these components/around them/

  components as well as an appropriate API for interacting with the container.
- **Security & Testability**. Having independent components corresponding to

>   virtualization components makes it easier to run fuzzing on individual

Reviewer: I would rephrase "Having independent components makes fuzz testing easy to apply to each individual package."

>   packages. The crate we want to focus on in the near future is

Reviewer: I would rephrase "We identified vm-virtio as the most important crate to be tested as it provides...."

  [vm-virtio](https://github.com/rust-vmm/vm-virtio) which provides a
  [virtio](https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=virtio)

>   device implementation. We want to keep a high standard in terms of testing

Reviewer: s/We/Generally, we/

Author: I woudn't use generally here because I think we should insist on a high quality for crates. Adding generally implies that we leave room for not-so-well tested components which is something that I would like to avoid (at least for the ones that are published to crates.io).

Reviewer: Fair enough!

  as these virtualization packages are going to be used in production by
  multiple projects. Each component is individually tested with a set of

>   common integration tests that are responsible for running the unit tests,

Reviewer: s/that are responsible/responsible/

>   computing the coverage, coding style and other lint checks. In the future

Reviewer: I would rephrase "...running unit tests and linters, checking coding style, and computing code coverage."

>   we will test the integration of the rust-vmm components by providing a

Reviewer: s/will/are planning to/

  reference VMM implementation.
- **Clean interface**. With shared crates you also have to think about an
  interface flexible enough to be used by multiple VMMs. We have multiple
  rounds of design reviews up to the point we are confident the interface is

>   clean and re-usable by other projects.

Reviewer: s/re-usable/can be consumed/

Author: I kept reusable (without the hyphen).

> ## rust-vmm components

Reviewer: Do we want to use "components" here instead of "crates"?

Keep in mind that this status might not always be up to date. For the latest
status please check individual repositories.

> **Components with PRs in fly**

Reviewer: What about "New crates"? Or "New components"?

Author: I would like to use something that suggests that these are empty crates and we are working towards merging something in them. Do you think new components is enough?

These are the empty repositories that have PRs waiting to be merged.
TODO

> **Components in Development:**

Reviewer: What about "Existing crates" or "Existing components", with a subsection for the ones published to crates.io?
I'm saying this because even crates that are published to crates.io can still be improved and then a new release comes to update crates.io.

Author: Sure, I'll re-phrase.

These are packages for which we have most of the code out there, but
there might be some missing requirements for publishing them on crates.io.

> For the full list of these requirements please check the

Reviewer: "For the requirement list refer to the"

[Publishing on crates.io - Requirements list](#publishing-on-crates.io---requirements-list)
section.

- [vm-memory](https://github.com/rust-vmm/vm-memory): abstractions over a
  Virtual Machine memory.

**Components Published on [crates.io](crates.io):**

These are production-ready components that can be pulled as dependencies
directly from crates.io.

- [kvm-bindings](https://crates.io/crates/kvm-bindings): Rust FFI bindings
  to KVM generated using [bindgen](https://crates.io/crates/bindgen).
- [kvm-ioctls](https://crates.io/crates/kvm-ioctls): Safe wrappers over the
  KVM API.

## How to contribute

### Adding a new virtualization component

If you have an idea about a component that could be a great fit for rust-vmm,
please open a
[review request issue](https://github.com/rust-vmm/community/issues/new?assignees=&labels=&template=repository-review-request.md&title=Repository%2FCrate+Review+Request).
The issue is a pitch for the component you want to add so in it you should
write details about the purpose of the repository, how will it be used and

> other information that you find relevant.

Reviewer: "The issue must provide the explanations and justifications behind the need for a new component, highlighting the purpose of this new repository, how....."

Relevant discussions about the proposal will happen on the same issue, and once

> we agree on the name (this part is hard!) and the design, an empty repository

Reviewer: s/we agree on/an agreement is reached about/

> is created in the rust-vmm GitHub organization.

Reviewer: s/is/will be/

To add functionality you will need to send Pull Requests (PRs) against the
newly created repository. While we do not expect the code to be perfect from
the first try, we do want to keep a high standard in terms of quality for all
existing repositories under rust-vmm.

### Become a member of rust-vmm

> 1. Submit a new [access request issue](https://github.com/rust-vmm/community/issues/new?assignees=andreeaflorescu%2C+jbryce%2C+sameo&labels=&template=access-request-to-rust-vmm-organization.md&title=Access+Request+to+rust-vmm).

Reviewer: We should remove "1."

## rust-vmm development

### Publishing on crates.io - Requirements list

> We consider crates published on crates.io to be production-ready, so to make

Reviewer: s/production-ready,/production ready./

> sure we keep a high security and quality standards, the following list of

Reviewer: I would rephrase "In order to maintain high security....."

requirements must be checked before having the first version published:

- **High level documentation**. This documentation should be added
   to `src/lib.rs` and it will end up being the home page of the crate on
   [docs.rs](https://docs.rs/). Here you should include information about what
   is the crate suppose to do, how can it be used in other projects as well as
   usage examples. Check out the
   [kvm-ioctls docs](https://docs.rs/kvm-ioctls/0.1.0/kvm_ioctls/) for an
   example.

> - **Documentation for the public interface**. Everything beloning to the public

Reviewer: s/beloning/belonging/

  methods should also have usage examples. The code examples are particularly
  useful because they are run when running the unit tests with `cargo test`. It
  is a nice way to ensure that your comments and examples don't are kept in
  sync with the code. To make sure we don't miss anything by mistake this rule
  can be enforced by using `#![deny(missing-docs)]` in `src/lib.rs`.
- **Unit Tests**.
- **Integration Tests**. All components will eventually be tested using the
  same common [Buildkite pipeline](https://buildkite.com/docs/pipelines). This
  is still WIP. The tests include style checks and other lint checks, a
  coverage test, builds with various configurations and on different
  architectures and operating systems.
- **README.md**
- **LICENSE**
- **CODEOWNERS**

### Versioning crates - WIP

This section is under construction.

### CI - WIP

We currently have the CI running on either [Travis](https://travis-ci.org/) or
[Buildkite](http://buildkite.com/). We are working towards having a central
repository for the CI so stay tuned for details on how we are testing the code.

This section is under construction.

> ### Fuzzing - WIP

Reviewer: "Fuzzing" or "Fuzz Testing"?

# Community

## What is rust-vmm

> rust-vmm is an open-source project enabling the community to build custom

Author: empowering? :D :D :D :D

Virtual Machine Monitors (VMMs) and Hypervisors. It offers a base of

> virtualization components that any project can use benefit from in order to

Reviewer: s/use//

quickly develop virtualization solutions while focusing on the key

> differentiators of their product rather than re-implementing common components

Author: Je ne sais pas. Should it?

> differentiators of their product rather than re-implementing common components

Reviewer 2: Yes.

like KVM wrappers, virtio devices and other VMM libraries.

The rust-vmm project is organized as a shared-effort, shared-ownership
open-source project that includes (so far) contributors from Alibaba, AWS,
Cloud Base, Crowdstrike, Intel, Google, Red Hat as well as individual
contributors.

Each virtualization component lives in a separated GitHub repository under the
rust-vmm organization. The code is organized such that one repository
corresponds to one
[Rust crate](https://doc.rust-lang.org/book/ch07-01-packages-and-crates-for-making-libraries-and-executables.html).

> ## Why rust-vmm

Reviewer 2: Sounds good to me. Linter should not scream because we want to be correct.

  components as well as an appropriate API for interacting with the container.
- **Security & Testability**. Having independent components makes fuzz testing
  easy to apply to each individual package. Our top priority is now
  [vm-virtio](https://github.com/rust-vmm/vm-virtio) as it provides a
  [virtio](https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=virtio)
  device implementation. We want to keep a high standard in terms of testing
  as these virtualization packages are going to be used in production by
  multiple projects. Each component is individually tested with a set of
  common integration tests responsible for running unit tests and linters
  (including coding style checks), and computing the coverage. In the future
  we are planning to test the integration of the rust-vmm components by
  providing a reference VMM implementation.
- **Clean interface**. With shared crates you also have to think about an
  interface flexible enough to be used by multiple VMMs. We have multiple
  rounds of design reviews up to the point we are confident the interface is
  clean and reusable by other projects.

> ## rust-vmm crates

Reviewer 2: So the whole sections below that one is a description of the crates statuses. I suggest we have 2 different sections:

* A description of the rust-vmm crates lifecycle: From initial crate request to then first PR merge and eventually making it to crates.io. And then also how code is getting pushed to crates.io. We should describe all of this in that section.

* An actual status of the current crates. I believe this could live in its own file, `crates.md` and be referenced here. The trick here is that either way it's often going to be outdated, and the only valid source for a given crate status is going to be the crate repo itself. So maybe we should not try to keep track of the crate statuses in a documentation file at all?

Reviewer 2: > Regarding your first bullet, I believe the section Adding a new virtualization component should give the information you are talking about. Are you suggesting to move that section here? I thought it makes more sense in the development section but I can also move it.
> 
Partially, yes. It does not mention the different steps a crate is expected to follow, and how a crate can "graduate" from one step to the next one.

> Regarding the status, I think it is important to give a high level description of crates, we don't have to have it up to date every time and it is also impossible as you already state it. We can also have it in a different file, I don't mind, but I think it gives readers the opportunity to quickly scroll through all the existing components without having to go to every single repository. If someone is interested in an in-depth description of a particular component then I'll assume they'll go and read about it on the component repository.

Then maybe we should have it in a more concise form, a table for example 

| Crate  | Description | Status |
| ------------- | ------------- | ----------|
|  [kvm-ioctls](https://crates.io/crates/kvm-ioctls) |  Safe wrappers over the KVM API  |  Published on crates.io |
| [vm-memory](https://github.com/rust-vmm/vm-memory) | rust-vmm memory model | Merged on rust-vmm |

Author: I find tables hard to read in text mode, but ok, sounds good.

Author: I take that back. Tables are absolutely hideous in markdown so if you don't feel strongly against having it with sections instead of tables I would leave it as is. I'll add some more details about the status and how we get from one stage to the other.

Reviewer 2: They're not nice in markdown, but much easier to read when rendered.

- [kvm-ioctls](https://crates.io/crates/kvm-ioctls): Safe wrappers over the
  KVM API.

## How to contribute

### Adding a new virtualization component

If you have an idea about a component that could be a great fit for rust-vmm,
please open a
[review request issue](https://github.com/rust-vmm/community/issues/new?assignees=&labels=&template=repository-review-request.md&title=Repository%2FCrate+Review+Request).
The issue must provide the explanations and justifications behind the need for
a new component, highlighting the purpose of this new repository, how will it
be used by other VMMs or Hypervisors, and other information that you find
relevant.

Discussions about the proposal will happen on the same issue, and once an
agreement is reached regarding the name (this part is hard!) and the design,
an empty repository will be created in the rust-vmm GitHub organization.

To add functionality you will need to send Pull Requests (PRs) against the
newly created repository. While we do not expect the code to be perfect from
the first try, we do want to keep a high standard in terms of quality for all
existing repositories under rust-vmm.

> ### Become a member of rust-vmm

Reviewer 2: I suggest we stop doing this. Or at least we explain what it means to become a member of the org and what it brings in terms of permissions, etc. And we make it clear that non members can contribute as well.

We consider crates published on crates.io to be production ready. In order to
maintain a high security and quality standard, the following list of
requirements must be checked before having the first version published:

- **High level documentation**. This documentation should be added
   to `src/lib.rs` and it will end up being the home page of the crate on
   [docs.rs](https://docs.rs/). Here you should include information about what
   is the crate suppose to do, how can it be used in other projects as well as
   usage examples. Check out the
   [kvm-ioctls docs](https://docs.rs/kvm-ioctls/0.1.0/kvm_ioctls/) for an
   example.
- **Documentation for the public interface**. Everything belonging to the
  public interface of the crate must be properly documented. Where applicable,
  the methods should also have usage examples. The code examples are
  particularly useful because they are run when running the unit tests with
  `cargo test`. It is a nice way to ensure that your comments and examples
  don't go out of sync with the code. To make sure we don't miss anything this
  rule is enforced by using `#![deny(missing-docs)]` in `src/lib.rs`.

> - **Unit Tests**.

Reviewer 2: Should we mention code coverage here?

Reviewer 2: I don't think we should. But we should clearly state that a PR which decreases the code coverage will not be merged.

- **Integration Tests**. All components will eventually be tested using the
  same common [Buildkite pipeline](https://buildkite.com/docs/pipelines). This
  is still WIP. The tests include style checks and other lint checks, a
  coverage test, builds with various configurations and on different
  architectures and operating systems.
- **README.md**

> - **LICENSE**

Reviewer 2: Here we should specify our preference, i.e. what we agreed at the Denver PTG and in the mailing list: We prefer dual licensing our crates with Apache 2.0 OR BSD 3-Clause and describe why. And that this dual licensing is obviously a preference for crates code that is written from scratch. We are not asking for re-licensing existing code.

# Community

> ## What is rust-vmm?

Author: I think it looks a bit weird with Is capitalize.

> ## What is rust-vmm?

Reviewer 3: I think this is defined in English. I found some example online apparently based on [The Chicago Manual of Style](https://www.chicagomanualofstyle.org/book/ed17/part2/ch08/toc.html):

> [...] if a two-letter word is acting as a noun, pronoun, adjective, or adverb, it must be capitalized. For example:
> 
> Go Tell it on the Mountain
> (wrong; "it" is a pronoun and should be capitalized)
> 
> When is a Spade a Spade?
> (wrong; "is" is a verb and should be capitalized)

rust-vmm is an open-source project empowering the community to build custom

> Virtual Machine Monitors (VMMs) and Hypervisors. It provides a set of

Author: Well @raduweiss you know I like to change my mind often. Maybe it can be used for hypervisors as well. We don't know and I don't want to be so restrictive here.

> Virtual Machine Monitors (VMMs) and Hypervisors. It provides a set of

Reviewer 3: Ok ok.

- **Clean interface**. With shared crates you also have to think about an
  interface flexible enough to be used by multiple VMMs. We have multiple
  rounds of design reviews up to the point we are confident the interface is
  clean and reusable by other projects.

## Status of rust-vmm Components

Keep in mind that this status might not always be up to date. For the latest
status please check individual repositories.

### Coming Soon

These are the empty repositories that have PRs waiting to be merged.

- [linux-loader](https://github.com/rust-vmm/linux-loader): parser and loader
  for vmlinux and bzImage images as well as some other helpers for kernel
  commandline.
- [virtio-bindings](https://github.com/rust-vmm/virtio-bindings): Rust FFI
  bindings to virtio kernel headers generated using
  [bindgen](https://crates.io/crates/bindgen).
- [vm-virtio](https://github.com/rust-vmm/vm-virtio/): virtio device trait and
  implementation for virtio primitives such as virtqueues and descriptor chain.
- [vmm-sys-utils](https://github.com/rust-vmm/vmm-sys-util/): a collection of
  modules providing helpers and utilities for building VMMs and Hypervisors.

> - [vmm-vcpu](https://github.com/rust-vmm/vmm-vcpu/): a hypervisor-agnostic

Reviewer 2: In that case, I think `hypervisor` is the right term.

there might be some missing requirements for publishing them on crates.io.
For the requirement list refer to the
[Publishing on crates.io - Requirements list](#publishing-on-crates.io---requirements-list)
section.

- [vm-memory](https://github.com/rust-vmm/vm-memory): abstractions over a
  Virtual Machine memory.

#### Components Published on [crates.io](crates.io)

These are production-ready components that can be pulled as dependencies
directly from crates.io.

- [kvm-bindings](https://crates.io/crates/kvm-bindings): Rust FFI bindings
  to KVM generated using [bindgen](https://crates.io/crates/bindgen).
- [kvm-ioctls](https://crates.io/crates/kvm-ioctls): Safe wrappers over the
  KVM API.

## Development

### Adding a New Virtualization Component

If you have an idea about a component that could be a great fit for rust-vmm,
please open a
[review request issue](https://github.com/rust-vmm/community/issues/new?assignees=&labels=&template=new-crate-request.md&title=Crate+Addition+Request).
The issue must provide the explanations and justifications behind the need for
a new component, highlighting the purpose of this new repository, how will it

> be used by other VMMs or Hypervisors, and other information that you find

Author: @aghecenco Let me first clarify the vmm vs hypervisor thing first. Some of the rust-vmm components can be used to create hypervisors as well, there's nobody saying you can only create VMMs. So that's why throughout the doc you'll see VMMs or Hypervisors. I am not trying to say you can use VMM and Hypervisors interchangeably, I hope that's not how it reads like.

@raduweiss do you mean I should use Virtual Machine Managers instead of VMMs?

> be used by other VMMs or Hypervisors, and other information that you find

Reviewer 3: No no, that was one of my two sarcastic comments.

relevant.

Discussions about the proposal will happen on the same issue, and once an
agreement is reached regarding the name (this part is hard!) and the design,
an empty repository will be created in the rust-vmm GitHub organization.

To add functionality you will need to send Pull Requests (PRs) against the
newly created repository. While we do not expect the code to be perfect from

> the first try, we do want to keep a high standard in terms of quality for all

Author: The requirements for merging the code are not the same as the requirements to publish it on crates.io. So I can't rephrase it as per @raduweiss suggestion.

existing repositories under rust-vmm.

### Publishing on crates.io - Requirements List

We consider crates published on crates.io to be production ready. In order to
maintain a high security and quality standard, the following list of
requirements must be checked before having the first version published:

- **High level documentation**. This documentation should be added
   to `src/lib.rs` and it will end up being the home page of the crate on
   [docs.rs](https://docs.rs/). Here you should include information about what
   is the crate suppose to do, how can it be used in other projects as well as
   usage examples. Check out the
   [kvm-ioctls docs](https://docs.rs/kvm-ioctls/0.1.0/kvm_ioctls/) for an
   example.
- **Documentation for the public interface**. Everything belonging to the
  public interface of the crate must be properly documented. Where applicable,
  the methods should also have usage examples. The code examples are

>   particularly useful because they are run when running the unit tests with

Reviewer 4: ```suggestion
  particularly useful because they are run as part of `cargo test`.
```

# Community

## What is rust-vmm?

rust-vmm is an open-source project empowering the community to build custom
Virtual Machine Monitors (VMMs) and Hypervisors. It provides a set of
virtualization components that any project can use to quickly develop
virtualization solutions while focusing on the key differentiators of their
product rather than re-implementing common components like KVM wrappers, virtio

> devices and other VMM libraries.

Reviewer 3: It's definitely not a blocker for a first version. I suggest you get more wring feedback once this document starts to get more user traffic, since at that point clarity will matter more.

The rust-vmm project is organized as a shared-effort, shared-ownership

> open-source project that includes (so far) contributors from Alibaba, AWS,

Author: Removed the dashes from shared-*

Cloud Base, Crowdstrike, Intel, Google, Red Hat as well as individual

> contributors.

Author: I'll do this in a following PR.

> contributors.

Reviewer 3: Sounds good.

Each virtualization component lives in a separated GitHub repository under the
rust-vmm organization. The code is organized such that one repository
corresponds to one
[Rust crate](https://doc.rust-lang.org/book/ch07-01-packages-and-crates-for-making-libraries-and-executables.html).

## Why rust-vmm?

- **Reduce code duplication**. The initial thought with rust-vmm was to create
  a place for sharing common virtualization components between two existing
  VMMs written in Rust:
  [CrosVM](https://chromium.googlesource.com/chromiumos/platform/crosvm/) and
  [Firecracker](https://github.com/firecracker-microvm/firecracker/). These
  two projects have similar code for calling KVM ioctls, managing the
  Virtual Machine memory, interacting with virtio devices and others. Instead
  of having these components live in each of the projects repositories, they
  can be shared through the rust-vmm project.
- **Faster development**. rust-vmm provides a base of virtualization components
  which are meant to be generic such that they can be consumed by other
  projects besides CrosVM and Firecracker. One example that is often mentioned

>   is building a container specific VMM. By doing so, the container VMM can

Reviewer 3: You can both be concise and give an example :)

  reuse most of the rust-vmm components and simply build the glue around them

>   as well as an appropriate API for interacting with the container.

Author: I was also thinking about saying something about cloud-hypervisor in this section, but still a container-specific VMM is often mentioned so I would still keep that reference even though it doesn't exit yet.

>   as well as an appropriate API for interacting with the container.

Reviewer 3: A specific example is much more powerful than an abstract case that's well understood only by a few people. I am working on the assumption that the readme aims to be informative for folks that are not yet deeply knowledgable about this space.

- **Security & Testability**. Having independent components makes fuzz testing
  easy to apply to each individual package. Our top priority is now
  [vm-virtio](https://github.com/rust-vmm/vm-virtio) as it provides a
  [virtio](https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=virtio)
  device implementation. We want to keep a high standard in terms of testing
  as these virtualization packages are going to be used in production by
  multiple projects. Each component is individually tested with a set of
  common integration tests responsible for running unit tests and linters
  (including coding style checks), and computing the coverage. In the future
  we are planning to test the integration of the rust-vmm components by

>   providing a reference VMM implementation.

Author: What do you mean? It will be used only for testing the integration of components. People will still need to have their own tests in their VMM. Should I mention that here?

>   providing a reference VMM implementation.

Reviewer 3: I think so, it's good to clearly say what rust-vmm will test and what it will not test.

- **Clean interface**. With shared crates you also have to think about an
  interface flexible enough to be used by multiple VMMs. We have multiple
  rounds of design reviews up to the point we are confident the interface is
  clean and reusable by other projects.

> ## Status of rust-vmm Components

Author: On what would you put the label though? You can see the status by looking at all repositories in rust-vmm and check what's in there which I find rather tedious. So as a quick starting point you can look at the status of crates in this doc (even if it is constantly changing).

> ## Status of rust-vmm Components

Reviewer 3: How about putting that that information in the repo description. It's neatly shown in the org repo list  [1]. There's even an API for that if you ever want to automated it.

[1] 
![image](https://user-images.githubusercontent.com/31901393/58718526-63ac7f80-83d6-11e9-9939-edf45dd8b0a7.png)
[2] https://developer.github.com/v3/repos/#list-organization-repositories

> ## Status of rust-vmm Components

Author: Let's do this in v2 of the document. I'll open an issue.

> ## Status of rust-vmm Components

Reviewer 3: Ack.

latest crate code is, the crate can be in one of 3 states:

1. `empty`: The crate repo exists but there is no code there yet. This
   typically means the crate was created after a community acknowledgement
   about its relevance to the project, but either no code has been submitted
   yet or the initial code base is being reviewed through a pull request.

1. `rust-vmm`: Once the initial code base has been merged, the crate may not
   be ready to be pushed to the [crates.io](https://crates.io) repository.
   In that state, the crate code is being improved and will be pushed by the
   rust-vmm maintainers once it meets all the
   [requirements](#publishing-on-cratesio---requirements-list) for being
   released to crates.io.

1. `crates.io`: When in that state, a crate is considered to be production
   ready. As such it can be consumed directly from the Rust canonical crate
   registry: [crates.io](https://crates.io). After meeting all the rust-vmm
   [requirements](#publishing-on-cratesio---requirements-list) for being
   published to crates.io, crates will move to the `crates.io` state. At that
   point, the crate repo in rust-vmm can be seen as a staging area until the

>    next push to crates.io. In other words, the production ready version of the

Reviewer 2: I'd rather keep that one.

   code lives in crates.io while development happens in the rust-vmm git repo.

> ### empty

Author: I don't have any ideas on how we could automate this. Sounds like a good idea, anyone has any idea how can this be achieved?

The readme is suppose to give people a hint about the components we have and doesn't need to be up to date all time. If people are interested in specific components they can go to the repositories and read more about them, check the status there and so on.

> ### empty

Reviewer 3: See my other comment on this.

> ### empty

Reviewer 2: There might be ways of doing this automatically, but I think this is material for a follow up PR.

- [vmm-sys-utils](https://github.com/rust-vmm/vmm-sys-util/): a collection of
  modules providing helpers and utilities for building VMMs and Hypervisors.
- [vmm-vcpu](https://github.com/rust-vmm/vmm-vcpu/): a hypervisor-agnostic
  abstraction for Virtual CPUs (vCPUs).

### rust-vmm

- [vm-memory](https://github.com/rust-vmm/vm-memory): abstractions over a
  Virtual Machine memory.

#### crates.io

- [kvm-bindings](https://crates.io/crates/kvm-bindings): Rust FFI bindings
  to KVM generated using [bindgen](https://crates.io/crates/bindgen).
- [kvm-ioctls](https://crates.io/crates/kvm-ioctls): Safe wrappers over the
  KVM API.

## Development

### Adding a New Virtualization Component

If you have an idea about a component that could be a great fit for rust-vmm,
please open a
[review request issue](https://github.com/rust-vmm/community/issues/new?assignees=&labels=&template=new-crate-request.md&title=Crate+Addition+Request).
The issue must provide the explanations and justifications behind the need for
a new component, highlighting the purpose of this new repository, how will it
be used by other VMMs or Hypervisors, and other information that you find
relevant.

> Discussions about the proposal will happen on the same issue, and once an

Author: @raduweiss what do you mean?

> Discussions about the proposal will happen on the same issue, and once an

Reviewer 3: You state "Discussions about the proposal" ... which is an issue ... "will happen on the same issue".

agreement is reached regarding the name (this part is hard!) and the design,
an empty repository will be created in the rust-vmm GitHub organization.

To add functionality you will need to send Pull Requests (PRs) against the
newly created repository. While we do not expect the code to be perfect from
the first try, we do want to keep a high standard in terms of quality for all
existing repositories under rust-vmm.

### Publishing on crates.io - Requirements List

We consider crates published on crates.io to be production ready. In order to
maintain a high security and quality standard, the following list of
requirements must be checked before having the first version published:

- **High level documentation**. This documentation should be added
   to `src/lib.rs` and it will end up being the home page of the crate on
   [docs.rs](https://docs.rs/). Here you should include information about what

>    is the crate suppose to do, how can it be used in other projects as well as

Reviewer 3: ```suggestion
   this crate does, how can it be used in other projects, as well as
```

   usage examples. Check out the
   [kvm-ioctls docs](https://docs.rs/kvm-ioctls/0.1.0/kvm_ioctls/) for an
   example.
- **Documentation for the public interface**. Everything belonging to the

>   public interface of the crate must be properly documented. Where applicable,

Reviewer 3: Maybe link some documentation about what constitutes the public interface of a rust crate?

Author: Everything that is exported hence can be used by other crates that take a dependency on it. To me that's obvious and it's also not necessarily rust specific. If it's not obvious for everyone I suppose I can also define that here.

Reviewer 3: Ack. Feel free to resolve (I don't see a button for it).

  the methods should also have usage examples. The code examples are
  particularly useful because they are run when running the unit tests with

>   `cargo test`. It is a nice way to ensure that your comments and examples

Reviewer 3: Link Rust documentation around this functionality or examples? Many languages don't have built-in doc example testing.

  don't go out of sync with the code. To make sure we don't miss anything this

>   rule is enforced by using `#![deny(missing-docs)]` in `src/lib.rs`.

Reviewer 3: Do your tests fail if a crate's `src/lib.rs` does not have that directive?

Author: No.

Reviewer 3: Then ...
s/enforced/enforceable/

> - **Unit Tests**

Reviewer 3: TODO?

- **Coverage**

> - **Integration Tests**. All components will eventually be tested using the

Reviewer 3: How is this an integration test? Does it integrate more crates?

Author: These are integration tests as much as [build tests](https://github.com/firecracker-microvm/firecracker/tree/master/tests/integration_tests/build) are integration tests in Firecracker. We will have true integration tests once we have a reference VMM implementation. In the meantime cloud-hypervisor and firecracker are the integration tests that I know of :))

Reviewer 3: Well, the error of our ways (calling build tests integration tests in Firecracker) needn't be perpetuated in other projects :)

  same common [Buildkite pipeline](https://buildkite.com/docs/pipelines). This
  is still WIP. The tests include style checks and other lint checks, a
  coverage test, builds with various configurations and on different
  architectures and operating systems.
- **README.md**
- **LICENSE**
- **CODEOWNERS**

### Versioning Crates - WIP

This section is under construction.

### CI - WIP

> We currently have the CI running on either [Travis](https://travis-ci.org/) or

Reviewer 3: Do you still Travis?

Author: Yap. In kvm-bindings for sure.

## Status of rust-vmm Components

Each rust-vmm crate lives in its own repository. Depending on where the
latest crate code is, the crate can be in one of 3 states:

1. `empty`: The crate repo exists but there is no code there yet. This
   typically means the crate was created after a community acknowledgement
   about its relevance to the project, but either no code has been submitted
   yet or the initial code base is being reviewed through a pull request.

1. `rust-vmm`: Once the initial code base has been merged, the crate may not
   be ready to be pushed to the [crates.io](https://crates.io) repository.
   In that state, the crate code is being improved and will be pushed by the
   rust-vmm maintainers once it meets all the
   [requirements](#publishing-on-cratesio---requirements-list) for being

>    released to crates.io.

Reviewer 2: This is fixed now.
